### PR TITLE
Tutorial System

### PR DESCRIPTION
**Pull request by Zonespace27**
# About the pull request
Adds a tutorial system to the game, accessible from the lobby screen. The tutorial system is entirely isolated from the main game, allowing players to get a curated experience to be taught the mechanics of SS13 or specific roles within CM. The tutorial system is fully capable of supporting a theoretically infinite amount of players at once, each getting their own instance.

See below video for an example of the in-dev "Marine - Basic" tutorial.

https://www.youtube.com/watch?v=aWEtd6EAZWk 

# Explain why it's good for the game
Teaching new players how to play the game has always been a tough bit for us, so why not add a full-on tutorial system to get people into the know?

# To-Do:
This list is alive and will change over time.
If you are interested in coding a tutorial, know that it's very easy and that most of the heavy lifting's done for you! You can find a [tutorial creation guide here](https://hackmd.io/@mRAdleXgRfmKqh97O8ixSA/BJQsmO8kT), and you can additionally contact me on discord at any time with questions.

Also, you can find an example tutorial [here](https://github.com/cmss13-devs/cmss13/pull/4442/files#diff-843b2f84360b9b932dfc960027992f2b5117667962bfa8da14f9a35f0179a926).

Backend:
- [x] TGUI
- [x] Add finished tutorials to save files
- [x] Communicate to players with little playtime
- [ ] Suppress combat logs and similar done in tutorials

SS13:
- [x] Basics
- [x] Intents

Marine:

- [x] Basics
- [x] Medical
- [ ] Weaponry
- [ ] Comtech - Basics
- [ ] Medic - Basics
- [ ] FTL - Basics
- [ ] Smartgunner - Basics
- [ ] Specialist - Demolitionist
- [ ] Specialist - Scout
- [ ] Specialist - Pyrotechnician
- [ ] Specialist - Grenadier
- [ ] Specialist - Sniper
- [ ] Squad Leader - Basics

Xenomorph:

- [ ] Basics
- [ ] Builder Caste - Basics

# Changelog
:cl:
add: Added a tutorial system for various roles (and just general information), find it in the lobby screen.
/:cl:

### CODE REVIEW

**Comment by Hopekz** [comment]
Good change. The text is easy to miss depending on how fast you read.

Sure you can slow it down but I think the best solution for this would be to add it to the chat under nice big font plus maybe incorporate the same solution that lobotomy corporation did by dropping notes of what was said in terms of spawned in papers that people can read at their own speed.

File: code/datums/tutorial/_tutorial.dm [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+/// The proc to begin doing everything related to the tutorial
+/datum/tutorial/proc/start_tutorial(mob/starting_mob)
+	SHOULD_CALL_PARENT(TRUE)
+
+	if(!starting_mob?.client)
+		return FALSE
+
+	ADD_TRAIT(starting_mob, TRAIT_IN_TUTORIAL, TRAIT_SOURCE_TUTORIAL)
+
+	tutorial_mob = starting_mob
```
Comment (spookydonut) [inline code/datums/tutorial/_tutorial.dm:51]: register on qdel, logout, client detached, ghosted, etc and trigger destroy

**Comment by Zonespace27** [comment]
> Sure you can slow it down but I think the best solution for this would be to add it to the chat under nice big font plus maybe incorporate the same solution that lobotomy corporation did by dropping notes of what was said in terms of spawned in papers that people can read at their own speed.


Currently, your objective is displayed in the status panel, see the documentation for update_objective()

File: code/datums/tutorial/ss13/basic_ss13.dm [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	RegisterSignal(tutorial_mob, COMSIG_LIVING_SPEAK, PROC_REF(on_speak))
+
+/datum/tutorial/ss13/basic/proc/on_speak(datum/source)
+	SIGNAL_HANDLER
+
+	UnregisterSignal(tutorial_mob, COMSIG_LIVING_SPEAK)
+	message_to_player("Excellent. To finish the tutorial, enter the nearby cryopod by <b>dragging your sprite</b> to the outlined <b>cryopod</b>.")
+	update_objective("Enter the cryopod by dragging your sprite to it.")
+
+	var/obj/structure/machinery/cryopod/tutorial/end_pod = new(loc_from_corner(0, 0)) // ZONENOTE: come back to this when save files are done so I can make this end properly
```
Comment (Zonespace27) [inline code/datums/tutorial/ss13/basic_ss13.dm:73]: reminder for self

**Comment by cm13-github** [comment]
This pull request has conflicts, please resolve those before we can evaluate the pull request.

**Comment by cm13-github** [comment]
Conflicts have been resolved. A maintainer will review the pull request shortly.

**Comment by Zonespace27** [comment]
I'll finish this soon

**Pull request closed without merging**